docrawl/docrawl_client.py

```diff
@@ -0,0 +1,319 @@
+        self.kv_redis.set(key=self._kv_redis_key_scanned_elements, value=elements)
+
+    def get_browser_scanned_elements(self):
+        self.kv_redis.get(key=self._kv_redis_key_scanned_elements)
+
+    def set_browser_screenshot(self, screenshot: str):
+        self.kv_redis.set(key=self._kv_redis_key_screenshot, value=screenshot)
+
+    def get_browser_screenshot(self):
+        self.kv_redis.get(key=self._kv_redis_key_screenshot)
+
+    def is_browser_active(self):
+        # TODO: finish later
+        pid = self.get_browser_meta_data()['browser']['pid']
+
+        return psutil.pid_exists(pid)
+
+    def _initialize_browser_metadata(self, driver, headless):
+        browser_meta_data = {
+            "browser": {"driver": driver, "headless": headless},
+            "function": {"name": "init_function", "input": None, "done": False}
+        }
+
+        self.set_browser_meta_data(browser_meta_data)
+
+    def _wait_until_page_is_loaded(self, timeout=20):
+        # Load spider_requests and spider_functions
+        try:
+            is_page_loaded = self.get_browser_meta_data()['request']['loaded']
+        except Exception as e:
+            docrawl_logger.error(f'Error while loading is_page_loaded: {e}')
+            is_page_loaded = True
+
+        # First check if page is loaded
+        timeout_start = time.time()
+        while not is_page_loaded and time.time() < timeout_start + timeout:
+            try:
+                is_page_loaded = self.get_browser_meta_data()['request']['loaded']
+            except:
+                is_page_loaded = False
+            time.sleep(0.5)
+            docrawl_logger.info('Page is still loading, waiting 0.5 sec ...')
+
+        if is_page_loaded:
+            docrawl_logger.success('Page loaded')
+        else:
+            docrawl_logger.error('Page was not loaded')
+
+    def _wait_until_function_is_done(self, timeout):
+        # Load spider_requests and spider_functions
+        try:
+            is_function_done = self.get_browser_meta_data()['function']['done']
+        except Exception as e:
+            docrawl_logger.error(f'Error while loading is_function_done: {e}')
+            is_function_done = True
+
+        # Then check if function is done
+        timeout_start = time.time()
+        while not is_function_done and time.time() < timeout_start + timeout:
+            try:
+                is_function_done = self.get_browser_meta_data()['function']['done']
+            except:
+                is_function_done = False
+            time.sleep(0.5)
+            docrawl_logger.info('Function is still running, waiting 0.5 sec ...')
+
+        if is_function_done:
+            docrawl_logger.success('Function finished')
+        else:
+            docrawl_logger.error('Function was not finished')
+
+    def _execute_function(self, function, function_input=None, timeout=30):
+        docrawl_logger.info(f'Running function {function} with input: {function_input}')
```

Comment:
"function" could be named "function_name" to not confuse with calling + it could be ENUM so there is a clear list of operations that are acccepted